src/scripting/LuaVM_Hooks.cpp

```diff
@@ -0,0 +1,330 @@
+#include "Scripting.h"
+
+#include <Image.h>
+#include <Pattern.h>
+
+#include <console/Console.h>
+
+struct REDScriptContext
+{
+};
+
+struct ScriptStack
+{
+    uint8_t* m_code;
+    uint8_t pad[0x28];
+    void* unk30;
+    void* unk38;
+    REDScriptContext* m_context;
+};
+static_assert(offsetof(ScriptStack, m_context) == 0x40);
+
+static TScriptCall** GetScriptCallArray()
+{
+    static uint8_t* pLocation = FindSignature({ 0x4C, 0x8D, 0x15, 0xCC, 0xCC, 0xCC, 0xCC, 0x48, 0x89, 0x42, 0x38, 0x49, 0x8B, 0xF8, 0x48, 0x8B, 0x02, 0x4C, 0x8D, 0x44, 0x24, 0x20, 0xC7 }) + 3;
+    static uintptr_t finalLocation = (uintptr_t)pLocation + 4 + *reinterpret_cast<uint32_t*>(pLocation);
+
+    return reinterpret_cast<TScriptCall**>(finalLocation);
+}
+
+void LuaVM::HookLog(REDScriptContext* apContext, ScriptStack* apStack, void*, void*)
+{
+    RED4ext::CString text("");
+    apStack->unk30 = nullptr;
+    apStack->unk38 = nullptr;
+    auto opcode = *(apStack->m_code++);
+    GetScriptCallArray()[opcode](apStack->m_context, apStack, &text, nullptr);
+    apStack->m_code++; // skip ParamEnd
+    
+    if (Options::Get().Console)
+        Console::Get().GameLog(text.c_str());
+
+    //auto& luavm = Get();
+    //if (!luavm.IsInitialized())
+    //    luavm.PostInitialize();
+}
+
+static const char* GetChannelStr(uint64_t hash)
+{
+    switch (hash)
+    {
+#define HASH_CASE(x) case RED4ext::FNV1a(x): return x
+        HASH_CASE("AI");
+        HASH_CASE("AICover");
+        HASH_CASE("ASSERT");
+        HASH_CASE("Damage");
+        HASH_CASE("DevelopmentManager");
+        HASH_CASE("Device");
+        HASH_CASE("Items");
+        HASH_CASE("ItemManager");
+        HASH_CASE("Puppet");
+        HASH_CASE("Scanner");
+        HASH_CASE("Stats");
+        HASH_CASE("StatPools");
+        HASH_CASE("Strike");
+        HASH_CASE("TargetManager");
+        HASH_CASE("Test");
+        HASH_CASE("UI");
+        HASH_CASE("Vehicles");
+#undef HASH_CASE
+    }
+    return nullptr;
+}
+
+void LuaVM::HookLogChannel(REDScriptContext* apContext, ScriptStack* apStack, void*, void*)
+{
+    uint8_t opcode;
+
+    uint64_t channel_hash = 0;
+    apStack->unk30 = nullptr;
+    apStack->unk38 = nullptr;
+    opcode = *(apStack->m_code++);
+    GetScriptCallArray()[opcode](apStack->m_context, apStack, &channel_hash, nullptr);
+
+    RED4ext::CString text("");
+    apStack->unk30 = nullptr;
+    apStack->unk38 = nullptr;
+    opcode = *(apStack->m_code++);
+    GetScriptCallArray()[opcode](apStack->m_context, apStack, &text, nullptr);
+
+    apStack->m_code++; // skip ParamEnd
+    
+    auto channel_str = GetChannelStr(channel_hash);
+    std::string channel = channel_str == nullptr
+        ? "?" + std::to_string(channel_hash)
+        : std::string(channel_str);
+    if (Options::Get().Console)
+        Console::Get().GameLog("[" + channel + "] " +text.c_str());
+    
+    
+    auto& luavm = Get();
+    if (!luavm.IsInitialized())
+        luavm.PostInitialize();
```

Comment:
Not sure about this, it might cause a race condition, would be safer to delay the call to the Update